File: meeting_rooms_2.cpp
// Time Complexity : O(n*logn) // Sorting + Traversal
// Space Complexity :  O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this

class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {

        // sort based on Start time        
        sort(intervals.begin(), intervals.end());
        
        // Maintain a min priority queue for tracking start and end times
        priority_queue<int, vector<int>, greater<int>> pq;
        
        int ret = 0;
        
        // Traverse based on start times
        for(const auto & m : intervals)
        {
            // Remove meetings finished
            while(pq.size() && m[0] >= pq.top())
                pq.pop();
            
            pq.push(m[1]);

            // Keep track of max rooms at any point   
            ret = max(int(ret), int(pq.size()));
        }
        
        return ret;
    }
};

File: meeting rooms 2.py
# TC: O(NlogN)
# SC: O(N)

class Solution:
    def minMeetingRooms(self, intervals) -> int:
        
        start = []
        end = []
        
        for i,j in intervals:
            start.append(i)
            end.append(j)
        start.sort()
        end.sort()
        
        room = 0
        e = 0
        s = 0
        while s<len(start):
            room += 1
            if start[s]>=end[e]:
                room -= 1 
                e += 1
            s += 1
        
        return room